Svc/ComQueue/ComQueue.cpp

```diff
@@ -0,0 +1,308 @@
+    this->m_lock.unLock();
+}
+
+
+// ----------------------------------------------------------------------
+// Private helper methods
+// ----------------------------------------------------------------------
+
+void ComQueue::enqueue(const FwIndexType queueNum, QueueType queueType, const U8* data, const FwSizeType size) {
+    // Enqueue the given message onto the matching queue. When no space is available then emit the queue overflow event,
+    // set the appropriate throttle, and move on. Will assert if passed a message for a depth 0 queue.
+    const FwSizeType expectedSize = (queueType == QueueType::COM_QUEUE) ? sizeof(Fw::ComBuffer) : sizeof(Fw::Buffer);
+    const FwIndexType portNum = queueNum - ((queueType == QueueType::COM_QUEUE) ? 0 : COM_PORT_COUNT);
+    FW_ASSERT(expectedSize == size, size, expectedSize);
+    FW_ASSERT(portNum >= 0, portNum);
+    Fw::SerializeStatus status = this->m_queues[queueNum].enqueue(data, size);
+    if (status == Fw::FW_SERIALIZE_NO_ROOM_LEFT && !this->m_throttle[queueNum]) {
+        this->log_WARNING_HI_QueueOverflow(queueType, portNum);
+        this->m_throttle[queueNum] = true;
+    }
+    // When the component is already in READY state process the queue to send out the next available message immediately
+    if (this->m_state == READY) {
+        this->processQueue();
+    }
+}
+
+void ComQueue::sendComBuffer(Fw::ComBuffer& comBuffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_comRetry = comBuffer;
+    this->comQueueSend_out(0, comBuffer, 0);
+    this->m_state = WAITING;
+}
+
+void ComQueue::sendBuffer(Fw::Buffer& buffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_lock.lock();
+    // Buffer ownership transferred to downstream component. Clear local reference.
+    this->m_bufferRetry = Fw::Buffer(nullptr, 0, 0);
+    this->m_lock.unLock();
+    this->buffQueueSend_out(0, buffer);
+    this->m_state = WAITING;
+}
+
+void ComQueue::retryQueue() {
+    // Check that we are in the appropriate state
+    FW_ASSERT(this->m_state == READY);
+
+    // Retry the last message based on type
+    if (this->m_lastIndex < COM_PORT_COUNT) {
+        this->sendComBuffer(this->m_comRetry);
+    } else {
+        // Ensure the retry buffer ownership has been returned before resending it.
+        this->m_lock.lock();
+        FW_ASSERT(this->m_bufferRetry.getData() != nullptr);
+        Fw::Buffer retry_temp = this->m_bufferRetry;
+        this->m_lock.unLock();
+        this->sendBuffer(retry_temp);
+    }
+}
+
+void ComQueue::processQueue() {
+    FwIndexType priorityIndex = 0;
+    FwIndexType sendPriority = 0;
+    // Check that we are in the appropriate state
+    FW_ASSERT(this->m_state == READY);
+
+    // Clean-up retry buffer when sending something new
+    this->m_lock.lock();
+    if (this->m_bufferRetry.getData() != nullptr) {
+        this->retryDeallocate_out(0, this->m_bufferRetry);
+        this->m_bufferRetry = Fw::Buffer(nullptr, 0, 0);
```

Comment:
Same question as above: Why clear the retry buffer here, when we may not be sending from it? Shouldn't we clear it if and only if we send from it?

Reply:
Here we are selecting a new item and are not retrying.  Thus the stored retry buffer is sent for final deallocation.  We clear the local reference as the data is now invalid.

Reply:
Theoretically we can reduce to this case (READY -> no retry -> deallocate) and the the retry case (RETRY -> clear -> callback).

Clearing on every buffer send means we'd clear twice. in this case.

Reply:
OK. But why are we deallocating the retry buffer here at all? Shouldn't we send it and let it be deallocated downstream? If we're deallocating the buffer here, then it means it will never be sent.

Reply:
The way I would expect it to work is that we send a buffer B, the send either succeeds or fails. If it succeeds, we never see B again. If it fails, then B comes back for retry and we retry by resending. I'm not sure why we would ever deallocate B here.

Reply:
Unless there's some sort of timeout and we give up and deliberately drop a buffer?

Reply:
I think it would help to see the retry algorithm written out in the SDD.

Reply:
Ownership is received from upstream.  It is transferred downstream and handed back for the purposes of retry.  If retry is not necessary it is deallocated.  Otherwise it is transfered downstream and handed-back again.